File: Data Structures & Algorithms/maximum-product-subarray/submission-4.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        mini = nums[0]
        maxi = nums[0]
        res = nums[0]
        for i in range(1, len(nums)):
            x = nums[i]
            maxic = max(x, maxi * nums[i], mini * nums[i])
            minic = min(x, mini * nums[i], maxi * nums[i])
            res = max(res, maxic, minic)
            mini = minic
            maxi = maxic
        return res

File: Data Structures & Algorithms/target-sum/submission-4.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        
        auto dp = [&](auto&& dp, int i, int sum) {
            if (i == nums.size()){
                if (sum == target) return 1;
                return 0;
            } 
            return dp(dp, i + 1, sum + nums[i]) + dp(dp, i + 1, sum - nums[i]);
        };
        return dp(dp, 0, 0);
    }
};
